src/dc_tools.rs

```diff
@@ -1066,137 +1082,70 @@ pub unsafe fn dc_get_abs_path(
-    if pathNfilename_abs.is_null() {
-        return 0;
+pub fn dc_get_filebytes(context: &Context, path: *const libc::c_char) -> uint64_t {
+    let path_abs = dc_get_abs_path_safe(context, as_path(path));
+    match fs::metadata(&path_abs) {
+        Ok(meta) => meta.len() as uint64_t,
+        Err(_err) => 0,
     }
-
-    let p = std::ffi::CStr::from_ptr(pathNfilename_abs)
-        .to_str()
-        .unwrap();
-    let filebytes = match fs::metadata(p) {
-        Ok(meta) => meta.len(),
-        Err(_err) => {
-            return 0;
-        }
-    };
-
-    free(pathNfilename_abs as *mut libc::c_void);
-    filebytes as uint64_t
 }
 
-#[allow(non_snake_case)]
-pub unsafe fn dc_delete_file(context: &Context, pathNfilename: *const libc::c_char) -> libc::c_int {
-    let mut success: libc::c_int = 0i32;
-    let pathNfilename_abs = dc_get_abs_path(context, pathNfilename);
-    if pathNfilename_abs.is_null() {
-        return 0;
-    }
-    let p = std::path::Path::new(
-        std::ffi::CStr::from_ptr(pathNfilename_abs)
-            .to_str()
-            .unwrap(),
-    );
-
-    let res = if p.is_file() {
-        fs::remove_file(p)
+pub fn dc_delete_file(context: &Context, path: *const libc::c_char) -> libc::c_int {
+    let path_abs = dc_get_abs_path_safe(context, as_path(path));
+    let res = if path_abs.is_file() {
+        fs::remove_file(path_abs)
     } else {
-        fs::remove_dir_all(p)
+        fs::remove_dir_all(path_abs)
     };
 
     match res {
-        Ok(_) => {
-            success = 1;
-        }
+        Ok(_) => 1,
         Err(_err) => {
-            warn!(context, 0, "Cannot delete \"{}\".", as_str(pathNfilename),);
+            warn!(context, 0, "Cannot delete \"{}\".", as_str(path));
+            0
         }
     }
-
-    free(pathNfilename_abs as *mut libc::c_void);
-    success
 }
 
-#[allow(non_snake_case)]
-pub unsafe fn dc_copy_file(
+pub fn dc_copy_file(
     context: &Context,
     src: *const libc::c_char,
     dest: *const libc::c_char,
 ) -> libc::c_int {
-    let mut success = 0;
-
-    let src_abs = dc_get_abs_path(context, src);
-    let dest_abs = dc_get_abs_path(context, dest);
-
-    if src_abs.is_null() || dest_abs.is_null() {
-        return 0;
-    }
-
-    let src_p = std::ffi::CStr::from_ptr(src_abs).to_str().unwrap();
-    let dest_p = std::ffi::CStr::from_ptr(dest_abs).to_str().unwrap();
-
-    match fs::copy(src_p, dest_p) {
-        Ok(_) => {
-            success = 1;
-        }
+    let src_abs = dc_get_abs_path_safe(context, as_path(src));
+    let dest_abs = dc_get_abs_path_safe(context, as_path(dest));
+    match fs::copy(&src_abs, &dest_abs) {
+        Ok(_) => 1,
         Err(_) => {
-            error!(context, 0, "Cannot copy \"{}\" to \"{}\".", src_p, dest_p,);
+            error!(
+                context,
+                0,
+                "Cannot copy \"{}\" to \"{}\".",
+                src_abs.to_string_lossy(),
+                dest_abs.to_string_lossy(),
```

Comment:
`Path::display()` is what should be used to make a printable path instead of `.to_string_lossy()`.